format/amf3/Reader.hx

```diff
@@ -25,81 +25,324 @@
+					sealedMemberNames:sealedMemberNames
+				}
+			);
+		}
+		else if( n & 7 == 7 )
+		{
+			// externalizable
+			isExternalizable = true;
+			className = readString();
+			trace(Tools.decode(className));  // TODO make registered class feature or use Type.resolveClass?
+		}
+		else
+		{
+			throw "Invalid object traits";
+		}
+
 		var h = new Map();
-		if (dyn) {
-			var s;
-			while ( true ) {
-				s = readString();
-				if (s == "") break;
-				h.set(s, read());
+
+		var ret = AObject( h );
+
+		// save new object in reference table
+		complexObjectsTable.push( ret );
+
+		if( !isExternalizable )
+		{
+			// parse sealed member values
+			for (j in 0...sealedMemberNames.length)
+				h.set(sealedMemberNames[j], read());
+
+			// parse any dynamic members (name-value pairs) until empty string name is found
+			if (dyn) {
+				var s;
+				while ( true ) {
+					s = Tools.decode(readString());
+					if (s == "") break;
+					h.set(s, read());
+				}
 			}
 		}
-		else {
-			var a = new Array();
-			for (j in 0...n)
-				a.push(readString());
-			for (j in 0...n)
-				h.set(a[j], read());
+		else
+		{
+			throw "Externalizable not supported";
 		}
-		return h;
+
+		return ret;
 	}
 	
-	function readMap(n : Int) {
+	function readMap() {
+		var n:Int = readInt();  // get header
+		if( n & 1 == 0 )
+		{
+			// reference previous array object
+			return complexObjectsTable[n >> 1];
+		}
+		n >>= 1;  // the rest of the header is the number of entries in the map
 		var h = new Map();
-		i.readByte();
+		var ret = AMap( h );
+		complexObjectsTable.push(ret);
+		i.readByte();  // skip weakKeys
+		//var weakKeys = i.readByte() != 0;
+		//if( weakKeys )
+		//	throw "Dictionary with weakKeys not supported";
 		for ( i in 0...n )
 			h.set(read(), read());
-		return h;
+		return ret;
 	}
 
-	function readArray(n : Int) {
+	function readArray() {
+		var n:Int = readInt();  // get header
+		if( n & 1 == 0 )
+		{
+			// reference previous array object
+			return complexObjectsTable[n >> 1];
+		}
+		n >>= 1;  // the rest of the header is the dense indexed array length
 		var a = new Array();
-		read();
+		var m = new Map<String, Value>();
+		var ret = AArray( a, m );
+		complexObjectsTable.push(ret);
+		var assocName:String = Tools.decode(readString());
+		while( assocName.length != 0 )
+		{
+			// got associative array element, get value
+			m[assocName] = read();
+
+			// next name
+			assocName = Tools.decode(readString());
+		}
 		for( i in 0...n )
 			a.push(read());
-		return a;
+		return ret;
 	}
-	
-	function readBytes(n : Int) {
+
+	function readIntVector()
+	{
+		var header:Int = readInt();
+		if( header & 1 == 0 )
+		{
+			// reference previous vector object
+			return complexObjectsTable[header >> 1];
+		}
+		var len = header >> 1;
+		var fixed = i.readByte() != 0;
+		var a:EitherType<Vector<Value>,Array<Value>>;
+		if( fixed )
+			a = new Vector(len);
+		else
+			a = new Array();
+
+		for(r in 0...len)
+		{
+			a[r] = AInt( i.readInt32() );
```

Comment:
For now, `readIntVector` can be used to read both vector-int-type and vector-uint-type but high uint values will overflow and go negative on certain targets. If we ever see an `Input.readUInt32()` we can use that here.